### PR TITLE
Typing callbacks

### PR DESCRIPTION
**Pull request by justusschock**
## What does this PR do?

Part of #5023 and #7037

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/callbacks/early_stopping.py [hunk 1]
```diff
@@ -118,31 +122,32 @@ def _validate_condition_metric(self, logs):
...
         self.patience = callback_state['patience']
 
-    def on_validation_end(self, trainer, pl_module):
+    def on_validation_end(self, trainer: 'Trainer', pl_module: 'LightningModule') -> None:
         from pytorch_lightning.trainer.states import TrainerState
         if trainer.state != TrainerState.FITTING or trainer.sanity_checking:
-            return
+            return None
 
         self._run_early_stopping_check(trainer)
```
Comment (ananthsub) [inline pytorch_lightning/callbacks/early_stopping.py:148]: `return None` is excessive - why not `return` only?

Comment (justusschock) [reply pytorch_lightning/callbacks/early_stopping.py:148]: In a couple of cases where there was `return` but the return type wasn't only `None`, but some optional return, mypy complained that a plain return wouldn't be sufficient but it had to be explicit, so I did it like this all the time :)

**Comment by Borda** [comment]
@justusschock how far are we, mind list what is missing so others may help/contribute? :rabbit:

File: pytorch_lightning/callbacks/pruning.py [hunk 2]
```diff
@@ -282,33 +290,37 @@ def apply_lottery_ticket_hypothesis(self):
...
+                return None
             dst.data = src.data.to(dst.device)
 
+        if self._original_layers is None:
+            raise RuntimeError
         for d in self._original_layers.values():
-            copy, names = d["data"], d["names"]
-            if self._resample_parameters and hasattr(copy, "reset_parameters"):
+            copy: Module = d['data']  # type: ignore
+            names: List[Tuple[int, str]] = d["names"]  # type: ignore
```
Comment (carmocca) [inline pytorch_lightning/callbacks/pruning.py:304]: Why ignore these?

**Review by carmocca — COMMENTED** [review]
Pruning review

File: pytorch_lightning/callbacks/pruning.py [hunk 3]
```diff
@@ -218,37 +223,40 @@ def __init__(
...
 
         IF use_global_unstructured, pruning_fn will be resolved into its associated ``PyTorch BasePruningMethod``
         ELSE, pruning_fn will be resolved into its function counterpart from `torch.nn.utils.prune`.
 
         """
+
+        pruning_func: Callable
         if self._use_global_unstructured:
-            pruning_fn = _PYTORCH_PRUNING_METHOD[pruning_fn]
+            pruning_func = _PYTORCH_PRUNING_METHOD[pruning_fn]
```
[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/callbacks/pruning.py:243]: We are allowing re-definition in master now. Did not check if this branch is updated

https://github.com/PyTorchLightning/pytorch-lightning/blob/14e6b46c6e643880e169013e0433a84cecdfe482/setup.cfg#L114

File: pytorch_lightning/callbacks/pruning.py [hunk 4]
```diff
@@ -261,14 +269,14 @@ def make_pruning_permanent(self, pl_module: LightningModule):
                     hook.remove(module)
                     del module._forward_pre_hooks[k]
 
-    def _restore_original_weights(self, module: nn.Module, orig_module: nn.Module, tensor_name: str):
+    def _restore_original_weights(self, module: nn.Module, orig_module: nn.Module, tensor_name: str) -> None:
         trained = getattr(module, tensor_name)
         orig = getattr(orig_module, tensor_name)
         if trained is None or orig is None:
-            return
+            return None
```
[the remark this answers is not in the document]
Comment (justusschock) [reply pytorch_lightning/callbacks/pruning.py:277]: Yeah, in some cases mypy complains if you return either None or something else that you aren't explicit about that so I chose to always do it like this.

File: pytorch_lightning/callbacks/pruning.py (hunk 2 above)
Comment (justusschock) [reply pytorch_lightning/callbacks/pruning.py:305]: the dict d is of type `Dict[str, Union[Module, List[Tuple[int,str]]]` here, and mypy complains since theoretically `d['data']` could also be of type `List[Tuple[int, str]]` and `d['names']` could also be of type Module

File: pytorch_lightning/callbacks/pruning.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (justusschock) [reply pytorch_lightning/callbacks/pruning.py:244]: yeah, that was the redefinition thing :) WIll update it

File: pytorch_lightning/callbacks/pruning.py [hunk 5]
```diff
@@ -371,28 +383,29 @@ def on_before_accelerator_backend_setup(self, trainer, pl_module: LightningModul
             for i, (module, name) in enumerate(self._parameters_to_prune):
                 id_ = id(module)
                 self._original_layers.setdefault(id_, {"data": deepcopy(module), "names": []})
-                self._original_layers[id_]["names"].append((i, name))
+                self._original_layers[id_]["names"].append((i, name))  # type: ignore
```
[the remark this answers is not in the document]
Comment (justusschock) [reply pytorch_lightning/callbacks/pruning.py:386]: it says 'Tensor' not callable which I don't understand at all. I think it might be, because the type of the object we append something to might be inferred from torch.nn.Module as `Union[Tensor, Module]`

**Comment by carmocca** [comment]
@justusschock Why close this? You could open PRs for separate callbacks, for example, pruning should be ready to go
